Path/IO.hs

```diff
@@ -314,11 +329,187 @@ listDir path = do
 -- exceptions as 'listDir'.
 
 listDirRecur :: (MonadIO m, MonadThrow m)
-  => Path b Dir        -- ^ Directory to list
+  => Path b Dir                          -- ^ Directory to list
   -> m ([Path Abs Dir], [Path Abs File]) -- ^ Sub-directories and files
-listDirRecur path = do
-  items <- listDir path
-  foldl' mappend items `liftM` mapM listDirRecur (fst items)
+listDirRecur = walkDirAccum Nothing (\_ d f -> return (d, f))
+
+-- | Similar to 'listDirRecur' but can use predicates to select the files and
+-- directories returned.
+
+listDirRecurWith
+  :: (MonadIO m, MonadThrow m)
+  => (Path Abs Dir -> m Bool)            -- ^ Dir match predicate
+  -> (Path Abs File -> m Bool)           -- ^ File match predicate
+  -> Path Abs Dir                        -- ^ Top dir to traverse
+  -> m ([Path Abs Dir], [Path Abs File]) -- ^ Matched subdirs and files
+listDirRecurWith dirPred filePred =
+  walkDirAccum Nothing (writeWith dirPred filePred)
+
+-- | Similar to 'listDirRecurWith' but prunes the matched directories
+-- i.e. it returns the matched dirs but does not traverse them.
+
+listDirRecurWithPruned
+    :: (MonadIO m, MonadThrow m)
+    => (Path Abs Dir -> m Bool)            -- ^ Dir match predicate
+    -> (Path Abs File -> m Bool)           -- ^ File match predicate
+    -> Path Abs Dir                        -- ^ Top dir to traverse
+    -> m ([Path Abs Dir], [Path Abs File]) -- ^ Matched subdirs and files
+listDirRecurWithPruned dirPred filePred =
+  walkDirAccum (Just $ descendExcluding dirPred) (writeWith dirPred filePred)
+
+----------------------------------------------------------------------------
+-- Directory Walk
+----------------------------------------------------------------------------
```

Comment:
Please preserve style of headers. Every part in import section should have corresponding header in code with exactly the same header, looking like this, e.g.:

```haskell
----------------------------------------------------------------------------
-- Walking directory trees
```

The line of `-` is 76 characters long.